### ISSUE TITLE
numpy->AoS->arrow roundtrips are ridiculously inefficient

### ISSUE THREAD

**Issue by teh-cmc** [rerun-io/rerun#2884]
Consider this minimal example that builds a batch of `Arrow3D`s in Python:

```python
# 1. Compute origins & vectors (SoA)
angles = np.arange(start=0, stop=tau, step=tau * 0.01)
origins = np.zeros(100 * 3).reshape((-1, 3))
vectors = np.column_stack([np.sin(angles), np.zeros(100), np.cos(angles)])

# 2. Build Arrow3D components (conversion to AoS)
arrows = [rrd.Arrow3D(origin=o, vector=v) for o, v in zip(origins, vectors)]

# 3. Serialize as pyarrow data (conversion back to SoA!!!)
serialized = rrc.Arrows3D(arrows)
```

What's happening here is a bit ridiculous:
1. We efficiently compute origins and vectors in numpy-land (i.e. SoA).
2. We build `Arrow3D` components (AoS) from our numpy arrays, leaving numpy-land and entering the _extremely slow_ native python land.
3. We go _back_ (!) to SoA by serializing our `Arrow3D` components back into invidivual arrays.

TL;DR: Step 2 does nothing but drastically slow down the entire process.

This problem is of course common to all languages we support (it's generally much more UX-friendly for users to deal with AoS data, and then we have to internally turn it into SoA for storage/transport); but it is amplified by orders of magnitude in Python, where leaving numpy-land has an unbearable cost on performance.

We need a friendly API that allows users to bypass step 2 and go directly from SoA numpy to SoA arrow.
Perhaps with the right set of overrides?

**Comment by teh-cmc** [comment]
We've decided to tackle the problem by splitting compound components into finer-grained components where it makes sense (i.e. things that are very likely to be logged in large batches and have an obvious natural way of being split into).

E.g.:
- The `Arrow3D(origin, vector)` gets split into the `Origin3D(Vec3D)` and `Vector(Vec3D)` components.
- The `Box3D(origin, half_extents)` gets split into become the `Origin3D(Vec3D)` and `HalfSize3D(Vec3D)` components.

The issue might crop up again in the future in a way where splitting things up is not applicable... but that should get us a _long_ way until then.

**Comment by teh-cmc** [comment]
With #2898, the example above can now be written as:
```python
# 1. Compute origins & vectors (SoA)
angles = np.arange(start=0, stop=tau, step=tau * 0.01)
origins = np.zeros(100 * 3).reshape((-1, 3))
vectors = np.column_stack([np.sin(angles), np.zeros(100), np.cos(angles)])

# 2 & 3. Serialize as pyarrow data (still SoA)
serialized = rrc.Arrows3D(vectors, origins)
```
